src/ex_color/model.py

```diff
@@ -25,11 +46,104 @@ def __init__(self):
             nn.Sigmoid(),  # Keep RGB values in [0,1]
         )
 
-    # TODO: Don't return latents from here: use hooks or callbacks instead
-    def forward(self, x: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
+        # Register forward hook to capture latents
+        self.encoder.register_forward_hook(self._capture_latents_hook)
+
+    def _capture_latents_hook(self, module, input, output):
+        """Hook to capture latent representations from the encoder."""
+        self.current_latents = output
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Forward pass returning only the reconstructed RGB values."""
         # Get our bottleneck representation
         bottleneck = self.encoder(x)
 
         # Decode back to RGB
         output = self.decoder(bottleneck)
-        return output, bottleneck
+        return output
+
+    def training_step(self, batch, batch_idx):
+        """Lightning training step."""
+        batch_data, batch_labels = batch
+
+        # Get current timeline state
+        current_state = self.timeline.state
+
+        # Forward pass
+        outputs = self(batch_data)
+        current_results = InferenceResult(outputs, self.current_latents)
+
+        # Calculate primary reconstruction loss
+        primary_loss = self.loss_criterion(batch_data, current_results).mean()
+        losses = {'recon': primary_loss.item()}
+        total_loss = primary_loss
+        zeros = torch.tensor(0.0, device=batch_data.device)
+
+        # Calculate regularizer losses
+        for regularizer in self.regularizers:
```

Comment:
It's cool that we maintain such fine-grained control over training. But I wonder: does Lightning have a more idiomatic way to use regularizers?